### PR TITLE
Added caching layer & refesher

### PR DESCRIPTION
**Pull request by nikcio**

### CODE REVIEW

File: src/Skybrud.Umbraco.BorgerDk/Caching/BorgerDkCachingService.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+using Umbraco.Cms.Core.Scoping;
+using Umbraco.Cms.Infrastructure.Persistence;
+using Umbraco.Extensions;
+
+namespace Skybrud.Umbraco.BorgerDk.Caching {
+    public class BorgerDkCachingService {
+        private readonly IMemoryCache cache;
+        private readonly IScopeProvider scopeProvider;
+        private readonly ILogger<BorgerDkCachingService> logger;
+        private readonly MemoryCacheEntryOptions cacheEntryOptions;
```
Comment (abjerner) [inline src/Skybrud.Umbraco.BorgerDk/Caching/BorgerDkCachingService.cs]: Private fields should be prefixed with an underscore (see my comment about `.editorconfig`).

File: src/Skybrud.Umbraco.BorgerDk/Caching/BorgerDkCachingService.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            this.logger = logger;
+            cacheEntryOptions = new MemoryCacheEntryOptions()
+                .SetAbsoluteExpiration(TimeSpan.FromDays(1));
+
+            RefreshCache();
+        }
+
+        public void CacheArticle(BorgerDkArticleDto article) {
+            cache.Set(article.Id, article, cacheEntryOptions);
+            CurrentlyCachedIds.Add(article.Id);
```
Comment (abjerner) [inline src/Skybrud.Umbraco.BorgerDk/Caching/BorgerDkCachingService.cs]: The implementation around `CurrentlyCachedIds` seems a bit fishy. I think ideally you should use a dictionary instead for storing the articles, and then you might not need to use the memory cache.

File: src/Skybrud.Umbraco.BorgerDk/BorgerDkService.cs [hunk 3]
```diff
@@ -72,23 +86,27 @@ public partial class BorgerDkService {
...
+                    _logger.LogError(ex, "Unable to get article from the database");
+                    throw new Exception("Unable to get article from the database", ex);
                 }
-
             }
-
         }
 
-        public void Import(BorgerDkArticle article) {
+        public void Import(BorgerDkArticle article, bool useCache = false) {
```
Comment (abjerner) [inline src/Skybrud.Umbraco.BorgerDk/BorgerDkService.cs]: Calling this method should always update the article, so the `useCache` parameter should be removed.

**Review by abjerner — CHANGES_REQUESTED** [review]
Hi @nikcio 

Please describe a bit more about the way you have structured the code. Anyways, I haven't tested your code yet, but I do have some requests for changes:

**Service Layer**
`BorgerDkService` is supposed to be the layer between where ever the articles are stored in our end (in this case in the database), and the rest of the package. The service shouldn't know about any caches, and should always return the realtime result from the database when queried. This is a general pattern we use for services.

In your PR, `BorgerDkService` accesses the new `BorgerDkCachingService` class, while it should be the other way around. So when articles are imported through the `BorgerDkService` class, it should send a broadcast about each article that has been updated, and the cache layer should then pick that up, and add or replace the articles in the cache layer. `BorgerDkCache` would also be a better name than `BorgerDkCachingService` as the *service* part in the name indicates a different purpose.

**.editorconfig**
While this wasn't part of the task, you've also moved the `editorconfig` file. We've made a few additions to our "global" `editorconfig` file since Jesper added it to the project. You can find the newest here:

https://github.com/limbo-works/.editorconfig/blob/main/.editorconfig

For one, it now makes Visual Studio suggest that private fields should be prefixed with an underscore. You have added private fields in a few of the files, but without the underscore prefix. This should be changed - then we can also avoid the `this.` keyword.

**Duplicate IDs**
Borger.dk has two sets of articles, as they have a Danish site (borger.dk) and an English site (lifeindenmark.borger.dk). As a result of this, we may experience colliding IDs, so the cache layer should use a unique key - eg. based on the domain and the article ID. I can't remember whether it also may be relevant to use the municipality code as part of the cache key.

**Comment by nikcio** [comment]
@abjerner I've tried to implement the changes you've mentioned. Now the caching layer has to be called by itself to access the cache at all. I've kept the Memory cache because it allows for TTL (Time to live) which makes sure that the cache refreshes sometimes. I'm pretty sure this will help minimize errors caused by the old cache.

About the Ids then the ids are created from what is currently used in the database so:
`string id = domain + "_" + municipality + "_" + articleId;`
So if you think this isn't enough it should be changed when creating records in the database too.

**Comment by abjerner** [comment]
@nikcio please at least do the effort of testing your code.

Due to circular dependencies in your code, Umbraco couldn't really boot. I'm not sure how much you've tested the cache refresher either. So rather than trying to debug your code and get it up and running, I ended up writing the cache layer on my own assuming that would be faster. You can see my commit here: 47684628552cffa35de80fcc65297366519cb936

If you're not sure how to test your code, or not sure how the different parts work, then please ask for help rather than not mentioning anything 🙏

**Pull request closed without merging**